File: Python/OneLang/One/Ast/Statements.py
from OneLangStdLib import *
import OneLang.One.Ast.Types as types
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.References as refs
import OneLang.One.Ast.Interfaces as ints

class Statement:
    def __init__(self):
        self.leading_trivia = None
        self.attributes = None

class IfStatement(Statement):
    def __init__(self, condition, then, else_):
        self.condition = condition
        self.then = then
        self.else_ = else_
        super().__init__()

class ReturnStatement(Statement):
    def __init__(self, expression):
        self.expression = expression
        super().__init__()

class ThrowStatement(Statement):
    def __init__(self, expression):
        self.expression = expression
        super().__init__()

class ExpressionStatement(Statement):
    def __init__(self, expression):
        self.expression = expression
        super().__init__()

class BreakStatement(Statement):
    def __init__(self):
        super().__init__()

class ContinueStatement(Statement):
    def __init__(self):
        super().__init__()

class UnsetStatement(Statement):
    def __init__(self, expression):
        self.expression = expression
        super().__init__()

class VariableDeclaration(Statement):
    def __init__(self, name, type, initializer):
        self.name = name
        self.type = type
        self.initializer = initializer
        self.references = []
        self.mutability = None
        super().__init__()
    
    def create_reference(self):
        return refs.VariableDeclarationReference(self)

class WhileStatement(Statement):
    def __init__(self, condition, body):
        self.condition = condition
        self.body = body
        super().__init__()

class DoStatement(Statement):
    def __init__(self, condition, body):
        self.condition = condition
        self.body = body
        super().__init__()

class ForeachVariable:
    def __init__(self, name):
        self.name = name
        self.type = None
        self.references = []
        self.mutability = None
    
    def create_reference(self):
        return refs.ForeachVariableReference(self)

class ForeachStatement(Statement):
    def __init__(self, item_var, items, body):
        self.item_var = item_var
        self.items = items
        self.body = body
        super().__init__()

class ForVariable:
    def __init__(self, name, type, initializer):
        self.name = name
        self.type = type
        self.initializer = initializer
        self.references = []
        self.mutability = None
    
    def create_reference(self):
        return refs.ForVariableReference(self)

class ForStatement(Statement):
    def __init__(self, item_var, condition, incrementor, body):
        self.item_var = item_var
        self.condition = condition
        self.incrementor = incrementor
        self.body = body
        super().__init__()

class CatchVariable:
    def __init__(self, name, type):
        self.name = name
        self.type = type
        self.references = []
        self.mutability = None
    
    def create_reference(self):
        return refs.CatchVariableReference(self)

class TryStatement(Statement):
    def __init__(self, try_body, catch_var, catch_body, finally_body):
        self.try_body = try_body
        self.catch_var = catch_var
        self.catch_body = catch_body
        self.finally_body = finally_body
        super().__init__()
        if self.catch_body == None and self.finally_body == None:
            raise Error("try without catch and finally is not allowed")

class Block:
    def __init__(self, statements):
        self.statements = statements

File: PHP/src/OneLang/One/Ast/Statements.php
<?php

namespace OneLang\One\Ast\Statements;

use OneLang\One\Ast\Types\IVariableWithInitializer;
use OneLang\One\Ast\Types\IVariable;
use OneLang\One\Ast\Types\IHasAttributesAndTrivia;
use OneLang\One\Ast\Types\IAstNode;
use OneLang\One\Ast\Types\MutabilityInfo;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\References\ForVariableReference;
use OneLang\One\Ast\References\ForeachVariableReference;
use OneLang\One\Ast\References\VariableDeclarationReference;
use OneLang\One\Ast\References\IReferencable;
use OneLang\One\Ast\References\Reference;
use OneLang\One\Ast\References\CatchVariableReference;
use OneLang\One\Ast\Interfaces\IType;

class Statement implements IHasAttributesAndTrivia, IAstNode {
    public $leadingTrivia;
    public $attributes;
    
    function __construct() {
        $this->leadingTrivia = null;
        $this->attributes = null;
    }
}

class IfStatement extends Statement {
    public $condition;
    public $then;
    public $else_;
    
    function __construct($condition, $then, $else_) {
        parent::__construct();
        $this->condition = $condition;
        $this->then = $then;
        $this->else_ = $else_;
    }
}

class ReturnStatement extends Statement {
    public $expression;
    
    function __construct($expression) {
        parent::__construct();
        $this->expression = $expression;
    }
}

class ThrowStatement extends Statement {
    public $expression;
    
    function __construct($expression) {
        parent::__construct();
        $this->expression = $expression;
    }
}

class ExpressionStatement extends Statement {
    public $expression;
    
    function __construct($expression) {
        parent::__construct();
        $this->expression = $expression;
    }
}

class BreakStatement extends Statement {
    
}

class ContinueStatement extends Statement {
    
}

class UnsetStatement extends Statement {
    public $expression;
    
    function __construct($expression) {
        parent::__construct();
        $this->expression = $expression;
    }
}

class VariableDeclaration extends Statement implements IVariableWithInitializer, IReferencable {
    public $name;
    public $type;
    public $initializer;
    public $references;
    public $mutability;
    
    function __construct($name, $type, $initializer) {
        parent::__construct();
        $this->name = $name;
        $this->type = $type;
        $this->initializer = $initializer;
        $this->references = array();
        $this->mutability = null;
    }
    
    function createReference() {
        return new VariableDeclarationReference($this);
    }
}

class WhileStatement extends Statement {
    public $condition;
    public $body;
    
    function __construct($condition, $body) {
        parent::__construct();
        $this->condition = $condition;
        $this->body = $body;
    }
}

class DoStatement extends Statement {
    public $condition;
    public $body;
    
    function __construct($condition, $body) {
        parent::__construct();
        $this->condition = $condition;
        $this->body = $body;
    }
}

class ForeachVariable implements IVariable, IReferencable {
    public $name;
    public $type;
    public $references;
    public $mutability;
    
    function __construct($name) {
        $this->name = $name;
        $this->type = null;
        $this->references = array();
        $this->mutability = null;
    }
    
    function createReference() {
        return new ForeachVariableReference($this);
    }
}

class ForeachStatement extends Statement {
    public $itemVar;
    public $items;
    public $body;
    
    function __construct($itemVar, $items, $body) {
        parent::__construct();
        $this->itemVar = $itemVar;
        $this->items = $items;
        $this->body = $body;
    }
}

class ForVariable implements IVariableWithInitializer, IReferencable {
    public $name;
    public $type;
    public $initializer;
    public $references;
    public $mutability;
    
    function __construct($name, $type, $initializer) {
        $this->name = $name;
        $this->type = $type;
        $this->initializer = $initializer;
        $this->references = array();
        $this->mutability = null;
    }
    
    function createReference() {
        return new ForVariableReference($this);
    }
}

class ForStatement extends Statement {
    public $itemVar;
    public $condition;
    public $incrementor;
    public $body;
    
    function __construct($itemVar, $condition, $incrementor, $body) {
        parent::__construct();
        $this->itemVar = $itemVar;
        $this->condition = $condition;
        $this->incrementor = $incrementor;
        $this->body = $body;
    }
}

class CatchVariable implements IVariable, IReferencable {
    public $name;
    public $type;
    public $references;
    public $mutability;
    
    function __construct($name, $type) {
        $this->name = $name;
        $this->type = $type;
        $this->references = array();
        $this->mutability = null;
    }
    
    function createReference() {
        return new CatchVariableReference($this);
    }
}

class TryStatement extends Statement {
    public $tryBody;
    public $catchVar;
    public $catchBody;
    public $finallyBody;
    
    function __construct($tryBody, $catchVar, $catchBody, $finallyBody) {
        parent::__construct();
        $this->tryBody = $tryBody;
        $this->catchVar = $catchVar;
        $this->catchBody = $catchBody;
        $this->finallyBody = $finallyBody;
        if ($this->catchBody === null && $this->finallyBody === null)
            throw new \OneLang\Core\Error("try without catch and finally is not allowed");
    }
}

class Block {
    public $statements;
    
    function __construct($statements) {
        $this->statements = $statements;
    }
}
